ngraph/core/src/validation_util.cpp

```diff
@@ -1194,3 +1196,375 @@ void ngraph::evaluate_nodes(std::map<RawNodeOutput, HostTensorPtr>& value_map,
         evaluator.evaluate(value);
     }
 }
+
+bool ngraph::could_propagate(const Output<Node>& output, NodeVector& order)
+{
+    bool status = true;
+
+    std::deque<shared_ptr<Node>> nodes_to_calculate = {output.get_node_shared_ptr()};
+    order.push_back(output.get_node_shared_ptr());
+
+    while (status && !nodes_to_calculate.empty())
+    {
+        auto current_node = nodes_to_calculate.front();
+        nodes_to_calculate.pop_front();
+
+        if (current_node->inputs().empty() && !is_type<op::Constant>(current_node))
+            status = false;
+        else if (!is_type<op::v0::ShapeOf>(current_node) && !is_type<op::v3::ShapeOf>(current_node))
```

Comment:
Look like we will definitely forget about to extend this list of operations when we will create next version of ShapeOf operation. Can we have some node input descriptor property like `m_is_relevant_to_value` ?

Reply:
Then we will forget to put this descriptor property for the new version of ShapeOf too :) I'm not sure what is the best approach here. Can we take this to the next PR?